### PR TITLE
Allow players to adjust the size of the ingame UI

### LINKED ISSUE #10382: Arbitrary zoom / custom resolution support
This outlines my plan for fixing our issues relating to zoom, custom resolutions, and HiDPI support in one fell swoop.  I've referred to it several times in the past, so I figure it is time to document this properly.  From my perspective this supersedes all the previous tickets, but of course others are free to argue against this.

**User interface**

Two new dropdowns are included in the display settings: UI resolution and World resolution (like UT3 and i'm sure other games have done).  They contain values for "Native", a selection of popular defaults with a fixed width and a height calculated from the aspect ratio, and a "custom" option that exposes text fields for entering a manual width.  These set the (fixed) size of the UI, and the default size/zoom of the world.

Players can use ctrl + scroll to arbitrarily zoom the world (but not the UI) between the minimum and maximum limits (explained below).  The `.` key (currently used for pixel doubling) is used to reset the world to the default zoom configured in the settings.

**Implementation**

The world rendering (everything that scales with the world, so this excludes selection decorations, target lines, etc) is rendered into a framebuffer that is configured a viewport size calculated from the zoom and default resolution, and the result is rendered using a quad covering the window viewport.  The UI rendering (everything else from the world, and then the widgets) is similarly rendered first into a framebuffer (maybe the same as the first, depending on the cost of reconfiguring a buffer vs switching to a new one), and then scaled to fill the screen.

The default "Native" value is calculated by the window / display size divided by the native pixel scale (queried from SDL).  The pixel scale will be 2 for retina displays under OSX, or between 0.75 and 2 (I think?) for the different DPI options under Windows.  We can calculate the pixel scale by comparing the resolution that we asked for with the output from [SDL_GetRendererOutputSize](https://wiki.libsdl.org/SDL_GetRendererOutputSize).  The minimum supported resolution is defined in mod.yaml (with defaults for our mods of 1024 x 700), and the maximum values are queried from the GPU (the smaller of `GL_MAX_RENDERBUFFER_SIZE` / `GL_MAX_TEXTURE_SIZE`).

To enable HiDPI rendering we need to set `SDL_WINDOW_ALLOW_HIGHDPI` when [creating the window](https://wiki.libsdl.org/SDL_CreateWindow), and under OSX we also need to set `NSHighResolutionCapable` in Info.plist.  I have no idea how DPI scaling works across the various linux WMs or whether they are supported by SDL.

The world is always rendered with a "zoom" of 1 into the framebuffer, so all of the fiddly code using Viewport.Zoom can be removed (yay!).  The shader-level hack for rendering the depth buffer can also be removed in favour of rendering the depth texture from the framebuffer.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29533233-arbitrary-zoom-custom-resolution-support?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Finally fixes #10382.

~Depends on #17581.~

### CODE REVIEW

**Comment by pchote** [comment]
Rebased.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 1]
```diff
@@ -837,6 +853,78 @@ static void ShowBattlefieldCameraDropdown(DropDownButtonWidget dropdown, Graphic
 			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 500, validSizes, setupItem);
 		}
 
+		static void RecalculateWidgetLayout(Widget w, bool insideScrollPanel = false)
+		{
+			// HACK: Recalculate the widget bounds to fit within the new effective window bounds
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: These hacks will go away once #16821 is ready.

**Review by Mailaender — DISMISSED** [review]
Works as promised.

**Comment by dragunoff** [comment]
The selection boxes and bars don't seem to render properly. Tested on Windows 7, changed UI scale in settings without changing the system DPI scale:
selection-box-scale

Or is that the expected result?

**Comment by pchote** [comment]
This is the expected result - the weird triangular bits in the selection boxes is a long standing bug in the line renderer that has always been there with HiDPI rendering.

**Comment by pchote** [comment]
Updated:
* Fixed the settings "reset" settings button putting the scale into an inconsistent state that broke ingame rendering
* Rebased on latest bleed.

**Review by teinarss — APPROVED** [review]

**Pull request merged**